### PR TITLE
ISPN-5743 + ISPN-5753 Soft-index file-store bugs

### PR DESCRIPTION
**Pull request by rvansa**
https://issues.jboss.org/browse/ISPN-5743
https://issues.jboss.org/browse/ISPN-5753

### CODE REVIEW

**Comment by rvansa** [comment]
Wait a sec, store.size() does not work properly for expired entries.

**Comment by rvansa** [comment]
Fixed store.size() behaviour, ready for review (please label).

**Comment by rvansa** [comment]
@gustavonalle Would you mind taking look on this since you've reviewed the original PR? These fixes should be applied to 7.2 and 8.0 branches too, when merging please apply that there too (or I will file separate PRs once this is reviewed).

Btw., in February you've asked me to speed up clean cache restart - I've implemented that ( https://github.com/rvansa/infinispan/tree/ISPN-5782 ), I'll file PR for that after this is merged.

**Comment by gustavonalle** [comment]
@rvansa this looks great. I will take a look on Mon.

File: persistence/soft-index/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreTest.java [hunk 1]
```diff
@@ -106,6 +113,55 @@ public void testStopStartAndMultipleWrites() {
       assertEquals("v2", store.load("k1").getValue());
    }
 
+   // test for ISPN-5743
+   public void testStopStartWithRemoves() {
```
Comment (gustavonalle) [inline persistence/soft-index/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreTest.java]: It seems this test (or maybe another from this class) is causing the suite to hang, please take a look at the stack trace https://gist.github.com/gustavonalle/72d470eb075ce6e0fbd9
I was able to reproduce quickly by running  ```mvn clean test``` in a loop.
Tested on master as well and it is not affected, ```SoftIndexFileStoreTest``` always passes

Comment (rvansa) [inline persistence/soft-index/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreTest.java]: If that's working on master, on what branch is it failing?

Comment (rvansa) [inline persistence/soft-index/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreTest.java]: In the stack trace, there's a missing thread named BCS-Compactor; is there any log with exception?

Comment (gustavonalle) [inline persistence/soft-index/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreTest.java]: Got now:
```
3426 [ERROR] Index: Error in indexer thread
java.lang.IllegalStateException: IndexRequest{key=foo, serializedKey=[B0x0301fe3e03666f6f, file=-1, offset=-1, prevFile=47, prevOffset=384, size=-1, type=DROPPED}
	at org.infinispan.persistence.sifs.Index$Segment.run(Index.java:260) [classes/:?]
Caused by: java.lang.IllegalStateException: Index cannot be outdated for segment updater thread
	at org.infinispan.persistence.sifs.IndexNode.copyWith(IndexNode.java:572) ~[classes/:?]
	at org.infinispan.persistence.sifs.IndexNode.setPosition(IndexNode.java:293) ~[classes/:?]
	at org.infinispan.persistence.sifs.Index$Segment.run(Index.java:257) ~[classes/:?]
Caused by: org.infinispan.persistence.sifs.IndexNode$IndexNodeOutdatedException: -1:-1 (1)
	at org.infinispan.persistence.sifs.IndexNode$LeafNode.getHeaderAndKey(IndexNode.java:918) ~[classes/:?]
	at org.infinispan.persistence.sifs.IndexNode$LeafNode.loadHeaderAndKey(IndexNode.java:905) ~[classes/:?]
	at org.infinispan.persistence.sifs.IndexNode.copyWith(IndexNode.java:570) ~[classes/:?]
	at org.infinispan.persistence.sifs.IndexNode.setPosition(IndexNode.java:293) ~[classes/:?]
	at org.infinispan.persistence.sifs.Index$Segment.run(Index.java:257) ~[classes/:?]
```

stack https://gist.github.com/gustavonalle/a4b7cc28b27062ba445c

**Comment by rvansa** [comment]
I've found few places which caused these exception, but stumbled upon a problem with Compactor I can't find an easy/elegant solution for yet.

**Comment by danberindei** [comment]
@rvansa The soft-index test suite is hanging.

**Comment by rvansa** [comment]
@danberindei I am aware of some problems, I'll have to work on this fix more...

**Comment by rvansa** [comment]
I'll close this for now to not delay CI when I now about some shortcomings.

**Pull request closed without merging**